python/kvikio/nvcomp_codec.py

```diff
@@ -226,20 +231,23 @@ def decode_batch(
             self._stream,
         )
 
+        # If all chunks are the same size, we can just return uncomp_chunks.
+        if is_equal_chunks and out is None:
+            return cp.asnumpy(uncomp_chunks) if is_host_buffer else uncomp_chunks
+
         res = []
         for i in range(num_chunks):
-            ret = uncomp_chunks[i]
-            if out is not None and out[i] is not None:
+            ret = uncomp_chunks[i, : uncomp_chunk_sizes[i].item()]
+            if out is None or out[i] is None:
+                res.append(cp.asnumpy(ret) if is_host_buffer else ret)
+            else:
                 o = ensure_contiguous_ndarray_like(out[i])
                 if hasattr(o, "__cuda_array_interface__"):
                     cp.copyto(o, ret.view(dtype=o.dtype), casting="no")
                 else:
+                    # TODO(akamenev): remove redundant H2H copy?
                     np.copyto(o, cp.asnumpy(ret.view(dtype=o.dtype)), casting="no")
```

Comment:
<!-- Polish comments are like a suggestion, where there is nothing necessarily wrong with the relevant content, there's just some ways to immediately improve the quality.  -->
**polish:** ‏addressing todo
```suggestion
                    cp.asnumpy(ret.view(dtype=o.dtype), out=o, stream=self._stream)
```

With a `self._stream.synchronize()` at the end of the loop.

Reply:
Done.